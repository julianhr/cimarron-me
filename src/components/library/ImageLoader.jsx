import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const Root = styled.figure`
  display: flex;
  align-items: center;
  background: ${props => props.theme.colors.background};
`

const Img = styled.img`
  opacity: 0;
  transition: opacity 100ms linear;
`

class ImageLoader extends React.PureComponent {
  static propTypes = {
    maxWidth: PropTypes.number.isRequired,
    maxHeight: PropTypes.number.isRequired,
    imgSrc: PropTypes.string.isRequired,
    styles: PropTypes.object,
  }

  static defaultProps = {
    styles: {},
  }

  state = {
    isImgLoaded: false,
  }

  refImg = React.createRef()

  handleOnLoad = () => {
    this.setState({ isImgLoaded: true })
  }

  render() {
    const { styles, maxWidth, maxHeight, imgSrc } = this.props
    let imgStyle = { willChange: 'opacity' }

    if (this.state.isImgLoaded) {
      imgStyle = {
        opacity: 1,
        width: this.refImg.current.width,
        height: this.refImg.current.height,
      }
    }

    return (
      <Root
        css={{ ...(styles.root || {}), width: maxWidth, height: maxHeight }}
      >
        <Img
          ref={this.refImg}
          src={imgSrc}
          css={{ ...(styles.img || {}), ...imgStyle }}
          onLoad={this.handleOnLoad}
        />
      </Root>
    )
  }
}

export default ImageLoader
