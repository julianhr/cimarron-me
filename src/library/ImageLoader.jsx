import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const Root = styled.figure`
  display: flex;
  align-items: center;
  background: ${props => props.theme.colors.background};
`

class ImageLoader extends React.PureComponent {
  static propTypes = {
    imgSrc: PropTypes.string.isRequired,
    maxHeight: PropTypes.number.isRequired,
    maxWidth: PropTypes.number.isRequired,
    shouldFadeIn: PropTypes.bool,
    styles: PropTypes.object,
  }

  static defaultProps = {
    styles: {},
    shouldFadeIn: true,
  }

  constructor() {
    super()
    this.handleOnLoad = this.handleOnLoad.bind(this)
  }

  state = {
    isImgLoaded: false,
  }

  handleOnLoad() {
    this.setState({ isImgLoaded: true })
  }

  getImgProps() {
    if (this.props.shouldFadeIn) {
      let imgStyle = {
        transition: 'opacity 80ms linear',
        willChange: 'opacity',
        opacity: 0,
      }
  
      if (this.state.isImgLoaded) {
        imgStyle = {
          transition: 'opacity 80ms linear',
          opacity: 1,
          width: '100%',
          height: 'auto',
        }
      }

      return {
        css: [this.props.styles.img, imgStyle],
        onLoad: this.handleOnLoad
      }
    } else {
      return { css: [this.props.styles.img] }
    }
  }

  render() {
    const { styles, maxWidth, maxHeight, imgSrc } = this.props

    return (
      <Root
        css={{ ...(styles.root || {}), width: maxWidth, height: maxHeight }}
      >
        <img
          src={imgSrc}
          {...this.getImgProps()}
        />
      </Root>
    )
  }
}

export default ImageLoader
