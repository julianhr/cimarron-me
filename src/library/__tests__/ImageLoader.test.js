import React from 'react'
import { render, mount } from 'enzyme'
import { create } from 'react-test-renderer'
import emotionSerializer from 'jest-emotion'

import ImageLoader from '../ImageLoader'
import MockApp from '~/__tests__/__mocks__/MockApp'
import imgSrc from '~/__tests__/__fixtures__/imgSrc'


const testProps = {
  maxWidth: 200,
  maxHeight: 100,
  imgSrc: imgSrc,
  shouldFadeIn: false,
  styles: {
    root: {
      color: 'green'
    },
    img: {
      color: 'yellow'
    }
  },
}

const renderApp = (props) => create(<MockApp><ImageLoader {...props} /></MockApp>)


describe('ImageLoader', () => {
  describe('props', () => {
    test('maxWidth is required', () => {
      const props = {...testProps}
      delete props.maxWidth
      expect(() => renderApp(props)).toThrow()
    })

    test('maxHeight is required', () => {
      const props = {...testProps}
      delete props.maxHeight
      expect(() => renderApp(props)).toThrow()
    })

    test('imgSrc is required', () => {
      const props = {...testProps}
      delete props.imgSrc
      expect(() => renderApp(props)).toThrow()
    })

    test('shouldFadeIn is not required', () => {
      const props = {...testProps}
      delete props.shouldFadeIn
      expect(() => renderApp(props)).not.toThrow()
    })

    test('styles is not required', () => {
      const props = {...testProps}
      delete props.styles
      expect(() => renderApp(props)).not.toThrow()
    })
  })

  describe('image fade in', () => {
    beforeAll(() => {
      expect.addSnapshotSerializer(emotionSerializer)
    })

    test('enabled', () => {
      const props = {...testProps}
      props.shouldFadeIn = true
      expect(renderApp(props)).toMatchSnapshot()
    })
  
    test('disabled', () => {
      const props = {...testProps}
      props.shouldFadeIn = false
      expect(renderApp(props)).toMatchSnapshot()
    })
  })

  describe('on image load', () => {
    let propsFadeIn

    beforeEach(() => {
      propsFadeIn = {...testProps}
      propsFadeIn.shouldFadeIn = true
    })

    it('should trigger onload function once loaded', () => {
      const oldHandle = ImageLoader.prototype.handleOnLoad
      const mockHandleOnLoad = jest.fn()
      ImageLoader.prototype.handleOnLoad = mockHandleOnLoad
  
      const wrapper = mount(<MockApp><ImageLoader {...propsFadeIn} /></MockApp>)
      wrapper.find('img').simulate('load')
      expect(mockHandleOnLoad).toBeCalledTimes(1)
      
      ImageLoader.prototype.handleOnLoad = oldHandle
      wrapper.unmount()
    })
    
    it('matches snapshot after image is loaded', () => {
      const wrapper = mount(<MockApp><ImageLoader {...propsFadeIn} /></MockApp>)

      wrapper.find('img').simulate('load')
      expect(wrapper).toMatchSnapshot()
      wrapper.unmount()
    })
  })
})
