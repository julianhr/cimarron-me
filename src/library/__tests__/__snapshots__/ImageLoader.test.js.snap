// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLoader image fade in disabled 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  color: green;
  width: 200px;
  height: 100px;
}

.emotion-0 {
  color: yellow;
}

<figure
  className="emotion-1 emotion-2"
>
  <img
    className="emotion-0"
    src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
  />
</figure>
`;

exports[`ImageLoader image fade in enabled 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  color: green;
  width: 200px;
  height: 100px;
}

.emotion-0 {
  color: yellow;
  -webkit-transition: opacity 80ms linear;
  transition: opacity 80ms linear;
  will-change: opacity;
  opacity: 0;
}

<figure
  className="emotion-1 emotion-2"
>
  <img
    className="emotion-0"
    onLoad={[Function]}
    src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
  />
</figure>
`;

exports[`ImageLoader on image load matches snapshot after image is loaded 1`] = `
Array [
  <ForwardRef(render)
    styles={[Function]}
  />,
  .emotion-3 {
  color: green;
  width: 200px;
  height: 100px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  color: green;
  width: 200px;
  height: 100px;
}

.emotion-0 {
  color: yellow;
  -webkit-transition: opacity 80ms linear;
  transition: opacity 80ms linear;
  opacity: 1;
  width: 0;
  height: 0;
}

<Root
    className="emotion-3"
  >
    <figure
      className="emotion-1 emotion-2"
    >
      <img
        className="emotion-0"
        onLoad={[Function]}
        src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
      />
    </figure>
  </Root>,
]
`;
